File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.py
import sys

sys.setrecursionlimit(10**6)

n, m, r = map(int, sys.stdin.readline().rstrip().split(" "))
graph = [[] for _ in range(n + 2)]
visited = [False] * (n + 1)
order = [0] * (n + 1)
cnt = 1


def dfs(node):
    global cnt

    visited[node] = True
    order[node] = cnt
    cnt += 1

    for next in sorted(graph[node], reverse=True):
        if not visited[next]:
            dfs(next)


for i in range(m):
    u, v = map(int, sys.stdin.readline().rstrip().split(" "))
    graph[u].append(v)
    graph[v].append(u)

dfs(r)

for i in range(1, n + 1):
    print(order[i])

File: 백준/Silver/24480. 알고리즘 수업 － 깊이 우선 탐색 2/알고리즘 수업 － 깊이 우선 탐색 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.StringTokenizer;

public class Main {

	static List<Integer>[] graph;
	static boolean[] visited;
	static int[] order;
	static int count = 1;

	public static void main(String[] args) throws IOException {

		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());

		int n = Integer.parseInt(st.nextToken());
		int m = Integer.parseInt(st.nextToken());
		int r = Integer.parseInt(st.nextToken());

		graph = new ArrayList[n + 1];
		for (int i = 1; i <= n; i++) {
			graph[i] = new ArrayList<>();
		}

		visited = new boolean[n + 1];
		order = new int[n + 1];

		for (int i = 0; i < m; i++) {
			st = new StringTokenizer(br.readLine(), " ");
			int u = Integer.parseInt(st.nextToken());
			int v = Integer.parseInt(st.nextToken());

			graph[u].add(v);
			graph[v].add(u);
		}

		for (int i = 1; i <= n; i++) {
			graph[i].sort(Collections.reverseOrder());
		}

		dfs(r);

		for (int i = 1; i <= n; i++) {
			System.out.println(order[i]);
		}
	}

	static void dfs(int node) {
		visited[node] = true;
		order[node] = count++;
		for(int next : graph[node]) {
			if (!visited[next]) {
				dfs(next);
			}
		}
	}
}
